bundles/org.openhab.core.io.console/src/main/java/org/openhab/core/io/console/internal/extension/UserConsoleCommandExtension.java

```diff
@@ -0,0 +1,197 @@
+                            User newUser = userRegistry.register(args[1], args[2], Set.of(args[3]));
+                            console.println(newUser.toString());
+                            console.println("User created.");
+                        } else {
+                            console.println("The user already exists.");
+                        }
+                    } else {
+                        console.printUsage(findUsage(SUBCMD_ADD));
+                    }
+                    break;
+                case SUBCMD_REMOVE:
+                    if (args.length == 2) {
+                        User user = userRegistry.get(args[1]);
+                        if (user != null) {
+                            userRegistry.remove(user.getName());
+                            console.println("User removed.");
+                        } else {
+                            console.println("User not found.");
+                        }
+                    } else {
+                        console.printUsage(findUsage(SUBCMD_REMOVE));
+                    }
+                    break;
+                case SUBCMD_CHANGEPASSWORD:
+                    if (args.length == 3) {
+                        User user = userRegistry.get(args[1]);
+                        if (user != null) {
+                            userRegistry.changePassword(user, args[2]);
+                            console.println("Password changed.");
+                        } else {
+                            console.println("User not found.");
+                        }
+                    } else {
+                        console.printUsage(findUsage(SUBCMD_CHANGEPASSWORD));
+                    }
+                    break;
+                case SUBCMD_LISTAPITOKENS:
+                    userRegistry.getAll().forEach(user -> {
+                        ManagedUser managedUser = (ManagedUser) user;
+                        if (managedUser.getApiTokens().size() > 0) {
+                            managedUser.getApiTokens().forEach(t -> {
+                                console.println("user=" + user.toString() + ", " + t.toString());
+                            });
+                        }
+                    });
+                    break;
+                case SUBCMD_ADDAPITOKEN:
+                    if (args.length == 4) {
+                        ManagedUser user = (ManagedUser) userRegistry.get(args[1]);
+                        if (user != null) {
+                            Optional<UserApiToken> userApiToken = user.getApiTokens().stream()
+                                    .filter(t -> args[2].equals(t.getName())).findAny();
+                            if (userApiToken.isEmpty()) {
+                                String tokenString = userRegistry.addUserApiToken(user, args[2], args[3]);
+                                // inform the user that the token will not be printed again, and they should save it?
```

Comment:
Is it even true the tokens cannot be retrieved?

I see there is a command to list them:

`openhab:users listApiTokens - lists the API keys for all users`

Reply:
> I think it would be helpful to print it. The UI also informs users of this.

I don't know, in the end I figured it's best to leave the command ouputs as no-frills as possible, in case the output is redirected to a file/piped to another command. The admin who masters these commands is supposed to know that it's the only opportunity to save the token.

> I see there is a command to list them

You only see the metadata like below, but not the actual token (and you can't even retrieve it from the JSON DB since it's not stored in plain text):
```
openhab> openhab:users listApiTokens
user=ysc (administrator), name=service1, createdTime=Mon Oct 19 20:17:40 CEST 2020, scope=admin
user=ysc (administrator), name=service2, createdTime=Mon Oct 19 20:17:46 CEST 2020, scope=admin
user=ysc (administrator), name=service3, createdTime=Mon Oct 19 20:17:50 CEST 2020, scope=admin
```

Reply:
Great! :+1: I didn't have time to compile and execute the command myself or figure out if it was hashed or plain text when I made the comment. Otherwise the next comment would have been to hash it. :wink:

Reply:
> I don't know, in the end I figured it's best to leave the command ouputs as no-frills as possible, in case the output is redirected to a file/piped to another command. The admin who masters these commands is supposed to know that it's the only opportunity to save the token.

Then the question has been answered and we should remove the comment. :-)

Reply:
I still wanted to hear opinions about this ;) Done now.